ananse/network.py

```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
-            # "factor_expression.rank.scale", 
-            # "target_expression.rank.scale",
-        ]
-        network = network[[c for c in network.columns if c not in exclude_cols]]
-        network = network.set_index("source_target")
-
-        bp = network.mean(axis=1)
-        bpd = pd.DataFrame(bp)
-        bpd = bpd.rename(columns={0: "binding"})
-        bpd["prob"] = minmax_scale(rankdata(bpd["binding"], method="dense"))
-
-        bpd.to_csv(outfile, sep="\t")
-
-    def run_network(self, binding, fin_expression, corrfiles, outfile):
-        # gene_bed="/home/qxu/.local/share/genomes/hg38/hg38_gffbed_piroteinCoding.bed"
-        # peak_bed="data/krt_enhancer.bed"
-        # pfmfile="../data/gimme.vertebrate.v5.1.pfm"
-        # binding="results/binding.predicted.h5"
-
-        # b=self.interaction.Interaction(genome=self.genome, gene_bed= self.gene_bed, pfmfile=self.pfmfile)
-        if not self.include_promoter and not self.include_enhancer:
-            logger.info("Create network")
-            self.create_expression_network(fin_expression, outfile)
-        elif self.include_promoter and not self.include_enhancer:
-            logger.info("Read data")
-            ddf = dd.read_csv(binding, sep="\t")[["factor", "enhancer", "binding"]]
-            tfdf, filter_bed = self.clear_peak(ddf)
-            
-            prom = self.get_promoter_dataframe(filter_bed)
+        tfs.rename(
+            columns={"target": "tf", "target_expression": "tf_expression"}, inplace=True
+        )
 
-            logger.info("Create network")
-            self.create_promoter_network(ddf, prom, fin_expression, outfile)
+        expression["key"] = 0
+        tfs["key"] = 0
 
-        else:
-            logger.info("Read data")
-            ddf = dd.read_csv(binding, sep="\t")[["factor", "enhancer", "binding"]]
+        network = expression.merge(tfs, how="outer")
+        network = network[["tf", "target", "tf_expression", "target_expression"]]
 
-            tfdf, filter_bed = self.clear_peak(ddf)
+        for col in ["tf_expression", "target_expression"]:
+            if rank:
+                network[col] = rankdata(network[col])
+            network[col] = minmax_scale(network[col])
 
-            prom = self.get_promoter_dataframe(filter_bed)
-            p = self.get_gene_dataframe(filter_bed)
+        network["tf_target"] = network["tf"] + "_" + network["target"]
+        network = network.set_index("tf_target").drop(columns=["tf", "target"])
+        logger.info("creating expression dataframe")
+        network = dd.from_pandas(network, npartitions=30)
+        # print(network)
+        return network
 
-            weight = self.distance_weight(self.include_promoter, self.include_enhancer)
+    def unique_enhancers(self, binding_fname, chrom=None):
+        p = re.compile("[:-]")
+        logger.info("reading enhancers")
+        enhancers = pd.read_table(binding_fname, usecols=["enhancer"])["enhancer"]
```

Comment:
This is **really** inefficient for large network files. However, the information is not present in an other file. Maybe the network should be stored in a different format, not in a flat text file. Could be a hdf5 file, with seperate tables per chromosome, and one table that stores all enhancers. Or something like that. You lose the benefit of flat text, but hopefully can store data more efficiently, and also enable more efficient reading.

Reply:
Another advantage of doing certain things per chromosome, is that we can do stuff in chunks parallel, or serial (in case of a machine with lower memory than cn45)

Reply:
One thing to consider is maybe sqlite r-tree. It's the perfect datastructure for this (linking proximalish peaks with tss) I think, and should be fast / dynamic. I implemented it for peaksql, and I could try a bit if it's (much) faster. The advantage is that we can put everything on disk, so the memory usage should be very low, and sqlite in general is fast. Big disadvantage is maintainability and perhaps portability between systems..